great_expectations/datasource/fluent/sql_datasource.py

```diff
@@ -883,6 +885,41 @@ def _create_batch_spec_kwargs(self) -> dict[str, Any]:
             "batch_identifiers": {},
         }
 
+    def _normalize_table_name_if_not_exists(
+        self,
+        engine: sqlalchemy.Engine,
+        inspector: sqlalchemy.Inspector,
+    ) -> None:
+        """
+        Oracle, DB2, and Snowflake store all case-insensitive object names in uppercase text.  In contrast, SQLAlchemy
+        considers all lowercase object names to be case-insensitive.  Hence, SQLAlchemy converts object name case during
+        schema-level communication, i.e. during table and index reflection.  If uppercase object names are used,
+        SQLAlchemy assumes they are case-sensitive and encloses the names with quotes.  This behavior will cause
+        mismatches against data dictionary data received from such DBMS, so unless identifier names have been created as
+        truly case-sensitive using quotes, e.g., "TestDb", all lowercase names should be used on SQLAlchemy side.
+
+        This method starts with assumption that table_name has correct casing.  If this assumption is incorrect, then
+        hypothesis is made that table_name is at least correct within quotation marks.  To validate this assumption, if
+        "self.table_name" exists (i.e., if inspection finds this name, identical, possibly short of quotation marks), in
+        database schema), then quoted version of supplied "self.table_name" is returned, and this property is updated.
+        """
+        table_exists = inspector.has_table(
+            table_name=self.table_name,
+            schema=self.schema_name,
+        )
+        if not table_exists:
+            typed_names: list[str | sqlalchemy.quoted_name] = inspector.get_table_names(
+                schema=self.schema_name
+            )
+            normalized_table_name_mapping: tuple[
+                str, str | sqlalchemy.quoted_name
+            ] = _verify_table_name_exists_and_get_normalized_typed_name_map(
+                name=self.table_name,
+                typed_names=typed_names,
+            )
+            if normalized_table_name_mapping:
+                self.table_name = normalized_table_name_mapping[1]
```

Comment:
You introduce a side-effect here. This updates the table name only when `test_connection` is called. I believe this is only called by us when interactively creating a new data asset but not when loading it from config. This feels like the wrong place.

2 Questions:
1. Can we sidestep this by always quoting the name and letting the db backend handle it? If it's case insensitive, it won't matter that sqlalchemy is lowercasing it or not. If it's case sensitive, the user needs to use the correct casing anyways.
2. When do we hit the issue of case insensitivity? Your docstring seems to indicate in the data dictionary. That's a different abstraction than this one and it feels strange to need to know about data dictionaries to update sql datasources. Can we instead normalize in the data dictionary abstraction?

Reply:
@billdirks These are good questions.  I like the idea of 1 above -- if I can make it work, that would be best.  I will try it and let you know what happens.  For 2 above, we hit it whenever the user decides to name their table in any way other than all-uppercase for Snowflake and other DBMS backends mentioned in the title.  The problem is the these DBMS backends and SQLAlchemy have different standards (SQLAlchemy standardizes on lowercase).  Let me see how 1 turns out, and then perhaps we can reason together about number 2.  Thank you.

Reply:
@billdirks For 1, it appears that your idea is working!  Thanks for pointing it out.  I would have to make sure, but if it is indeed the case, if we insert

```
table_name = sqlalchemy.quoted_name(value=self.table_name, quote=True)
```
just above the call to `inspector.has_table()` and use `table_name` instead of `self.table_name` (so we would not modify our instance variable), then this would fit well with our abstraction?  Thanks!

Reply:
This is more in line with what I'm thinking of. I'd rather keep `self.table_name` immutable so some code path does/does not work conditionally on calling this function first.
I wonder if, in the constructor, we could do `self.table_name = sqlalchemy.quoted_name(value=passed_in_table_name, quote=True)` and then we wouldn't have to worry about it here at all?

Reply:
@billdirks I am happy to do that -- as long as we are OK with the passed in `self.table_name` become quoted (as opposed to what the user originally specified).  I can experiment with it.  I will update you once I know more.  Thanks.